pkg/sql/exec/coalescer.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2018 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package exec
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/sql/exec/types"
+)
+
+// coalescerOp consumes the input operator and coalesces the resulting batches
+// to return full batches of ColBatchSize.
+type coalescerOp struct {
+	input      Operator
+	inputTypes []types.T
+
+	group  ColBatch
+	buffer ColBatch
+}
+
+var _ Operator = &coalescerOp{}
+
+// NewCoalescerOp creates a new coalescer operator on the given input operator
+// with the given column types.
+func NewCoalescerOp(input Operator, colTypes []types.T) Operator {
+	return &coalescerOp{
+		input:      input,
+		inputTypes: colTypes,
+		buffer:     NewMemBatch(colTypes),
+	}
+}
+
+func (p *coalescerOp) Init() {
+	p.input.Init()
+}
+
+func (p *coalescerOp) Next() ColBatch {
+	p.group = p.buffer
+	p.buffer = NewMemBatch(p.inputTypes)
+
+	for p.group.Length() < ColBatchSize {
+		leftover := ColBatchSize - p.group.Length()
+		batch := p.input.Next()
+		batchSize := batch.Length()
+
+		if batchSize == 0 {
+			break
+		}
+
+		sel := batch.Selection()
+
+		if batchSize <= leftover {
```

Comment:
This amount of duplication seems egregious - can you remove it? Since we're going column by column and not tuple by tuple, we can embed the `if sel` and `if batchSize <= leftover` within the for-each-type loop without incurring performance penalties.

Reply:
Done.